src/Symfony/Component/Messenger/Middleware/SecurityMiddleware.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Middleware;
+
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface;
+use Symfony\Component\Security\Core\Exception\AccessDeniedException;
+
+/**
+ * @author Sebastiaan Stok <s.stok@rollerscapes.net>
+ */
+final class SecurityMiddleware implements MiddlewareInterface
+{
+    private $authorizationChecker;
+
+    public function __construct(AuthorizationCheckerInterface $authorizationChecker)
+    {
+        $this->authorizationChecker = $authorizationChecker;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function handle(Envelope $envelope, StackInterface $stack): Envelope
+    {
+        $message = $envelope->getMessage();
+
+        if (!$this->authorizationChecker->isGranted([], $message)) {
```

Comment:
what is the point of `[]`? what about introducing stamps here that can be attached to the Envelope with a ROLE and then checked here?

Reply:
Great suggestion 👍 can you gave a more detailed explanation of this use-case?

Reply:
Now that I think about this, shouldn't this be a Ui concern? Eg. the controller checks if you are authorized to even dispatch the message? Why would you need this?

Reply:
I'm not using the Async handler but using the MessageBus system as a CQRS system. Whether you handle a query or dispatch a Command, the source of the dispatch can be anything from a Controller action to a Process Manager. While the initial message might have been allowed the secondary message might not.

I might be missing something 😅 I properly am, I don't have to many experience with CQRS and these cross cutting concerns are always a bit difficult to determine when to execute them.

_Because I already had the source code I didn't bother to open an RFC issue first, otherwise I definitely would have done this_ 👍

Reply:
I am not an expert either but I don't think this is a concern of the bus. I think this should be checked before doing the dispatch. Let's wait for someone with more experience.

Reply:
I just remembered Prooph had something similar 🤔 https://github.com/prooph/service-bus/tree/master/src/Plugin/Guard

Reply:
By the way, remember that if there's no attribute and no `Voter` that `supports()` the `subject` message, **all** voters will abstain by default, leading `isGranted()` to return `false`, so I don't know if it's active by default, but it should be used only if at least one voter supports handling messages.

Reply:
Maybe solution is to use a special `SecurityMessageInterface`?

Reply:
The voters would be access to the message, so they could read any stamps set on it, right?
Then no need for any attributes indeed, as the message can already embed anything voters could need to decide, I suppose. We could use attributes to pass on something that doesn't belong to the message itself - ie make them configurable as constructor arguments!

Reply:
Would you suggest to pass the Envelope as the subject of `isGranted` instead of the Message? Or pass the Envelelope (or it's stamps) as arguments? 🤔

Reply:
hum, it's true that's what I implicitly suggested... but maybe not a good idea if we consider that voters shouldn't be aware of the transport layer. No strong opinion...

Reply:
I think using `[]` is a bad idea : it means any voters extending the [Voter](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php) class will always abstain. I think we should either use a sting constant (`handle` ?) or require the user to configure it with stamps.

Reply:
I gave it some thought, adding Security attributes Stamp would properly make sense.

Secondly I thought about passing the authentication token, but I'm not entire sure if it would make sense, this would use the `AccessDecisionManager` instead. Serializing is already possible as the Token is stored in the session.